File: client/rust/src/deserialize.rs
use crate::error::{DeserializationError, DeserializationResult, Exists};
use crate::data::*;
use crate::twitter1;

use std::convert::TryFrom;
use chrono::{DateTime, NaiveDateTime, Duration, Utc};

pub(super) trait Deserialize<T> where T: Sized {
    fn des(self) -> DeserializationResult<T>;
}

impl<T,S> Deserialize<Option<S>> for Option<T> where T: Deserialize<S> {
    fn des(self) -> DeserializationResult<Option<S>> {
        self.map(T::des).map_or(Ok(None), |x| x.map(Some))
    }
}

impl<T,S> Deserialize<Box<S>> for Box<T> where T: Deserialize<S> {
    fn des(self) -> DeserializationResult<Box<S>> {
        Ok(Box::new((*self).des()?))
    }
}

impl<T,S> Deserialize<Vec<S>> for Vec<T> where T: Deserialize<S> {
    fn des(self) -> DeserializationResult<Vec<S>> {
        self.into_iter().map(T::des).collect()
    }
}

impl Deserialize<DateTime<Utc>> for i64 {
    fn des(self) -> DeserializationResult<DateTime<Utc>> {
        Ok(DateTime::from_utc(NaiveDateTime::from_timestamp(self, 0), Utc))
    }
}

impl Deserialize<DateTime<Utc>> for u64 {
    fn des(self) -> DeserializationResult<DateTime<Utc>> {
        match i64::try_from(self) {
            Ok(x)  => x.des(),
            Err(_) => Err(DeserializationError::FieldOutOfRange)
        }
    }
}

impl Deserialize<Handle> for String {
    fn des(self) -> DeserializationResult<Handle> {
        Ok(Handle{
            name_only: self,
        })
    }
}

impl Deserialize<Indices> for twitter1::Indices {
    fn des(self) -> DeserializationResult<Indices> {
        Ok(self.start as usize .. self.end as usize)
    }
}

impl Deserialize<Tweet> for twitter1::Tweet {
    fn des(self) -> DeserializationResult<Tweet> {
        Ok(Tweet{
            id: self.id,
            created_at: self.created_at.des()?,
            raw_text: self.text,
            text_display_range: self.text_display_range.exists()?.des()?,
            truncated: self.truncated,
            source: self.source,
            user: Box::new(self.user.exists()?.des()?),
            replied_to: self.replied_tweet.des()?,
            quoted: self.quoted_tweet.des()?,
            retweeted: self.retweeted_tweet.des()?,
            quotes: self.quote_count,
            replies: self.reply_count,
            retweets: self.retweet_count,
            likes: self.favorite_count,
            you_liked: self.favorited,
            you_retweeted: self.retweeted,
            your_retweet_id: self.current_user_retweet_id.map(u64::from),
            hashtags: self.hashtags.des()?,
            urls: self.urls.des()?,
            mentions: self.mentions.des()?,
            symbols: self.symbols.des()?,
            media: self.media.des()?,
            polls: self.polls.des()?,
            sensitive: self.possibly_sensitive,
            filter_level: self.filter_level,
            lang: self.lang,
            withheld_copyright: self.withheld_copyright,
            withheld_countries: self.withheld_countries,
            withheld_scope: self.withheld_scope,
        })
    }
}

impl Deserialize<Vec<Tweet>> for twitter1::Tweets {
    fn des(self) -> DeserializationResult<Vec<Tweet>> {
        self.tweets.into_iter()
            .map(twitter1::Tweet::des)
            .collect()
    }
}

impl Deserialize<ReplyData> for twitter1::tweet::ReplyData {
    fn des(self) -> DeserializationResult<ReplyData> {
        Ok(ReplyData{
            tweet_id: self.reply_to_tweet_id,
            user_id: self.reply_to_user_id,
            user_handle: self.reply_to_user_handle.des()?,
        })
    }
}

impl Deserialize<User> for twitter1::User {
    fn des(self) -> DeserializationResult<User> {
        Ok(User{
            id: self.id,
            handle: self.handle.des()?,
            name: self.display_name,
            created_at: self.created_at.des()?,
            bio: self.bio,
            url: self.url,
            location: self.location,
            protected: self.protected,
            verified: self.verified,
            follower_count: self.follower_count,
            following_count: self.following_count,
            listed_count: self.listed_count,
            favorites_count: self.favorites_count,
            statuses_count: self.statuses_count,
            profile_banner: self.profile_banner,
            profile_image: self.profile_image,
            default_profile: self.default_profile,
            default_profile_image: self.default_profile_image,
            withheld_countries: self.withheld_countries,
            withheld_scope: self.withheld_scope,
            urls: self.url_urls.des()?,
            bio_urls: self.bio_urls.des()?,
        })
    }
}

impl Deserialize<URL> for twitter1::Url {
    fn des(self) -> DeserializationResult<URL> {
        Ok(URL{
            indices: self.indices.exists()?.des()?,
            twitter_url: self.twitter_url,
            display_url: self.display_url,
            expanded_url: self.expanded_url,
        })
    }
}

impl Deserialize<Symbol> for twitter1::Symbol {
    fn des(self) -> DeserializationResult<Symbol> {
        Ok(Symbol{
            indices: self.indices.exists()?.des()?,
            text: self.text,
        })
    }
}

impl Deserialize<Mention> for twitter1::Mention {
    fn des(self) -> DeserializationResult<Mention> {
        Ok(Mention{
            indices: self.indices.exists()?.des()?,
            user_id: self.user_id,
            user_handle: self.handle.des()?,
            user_name: self.display_name,
        })
    }
}

impl Deserialize<Media> for twitter1::Media {
    fn des(self) -> DeserializationResult<Media> {
        Ok(Media{
            url: self.url.exists()?.des()?,
            id: self.id,
            media_type: self.r#type,
            media_url: self.media_url,
            alt: self.alt,
            source_tweet_id: self.source_tweet_id.map(u64::from),
            thumb: self.thumb.exists()?.des()?,
            small: self.small.exists()?.des()?,
            medium: self.medium.exists()?.des()?,
            large: self.large.exists()?.des()?,
        })
    }
}

impl Deserialize<MediaSize> for twitter1::media::Size {
    fn des(self) -> DeserializationResult<MediaSize> {
        Ok(MediaSize{
            width: self.width,
            height: self.height,
            resize: self.resize,
        })
    }
}

impl Deserialize<Poll> for twitter1::Poll {
    fn des(self) -> DeserializationResult<Poll> {
        Ok(Poll{
            end_time: self.end_time.des()?,
            duration: Duration::minutes(self.duration_minutes as i64),
            options: self.options.des()?,
        })
    }
}

impl Deserialize<PollOption> for twitter1::poll::Option {
    fn des(self) -> DeserializationResult<PollOption> {
        Ok(PollOption{
            position: self.position as usize,
            text: self.text,
        })
    }
}

File: client/go/deserialize.go
package goldcrest

import (
  pb "github.com/pantonshire/goldcrest/protocol"
  "time"
)

func desTimeline(msg *pb.Tweets) []Tweet {
  if msg == nil {
    return nil
  }
  tweets := make([]Tweet, len(msg.Tweets))
  for i, tweetMsg := range msg.Tweets {
    tweets[i] = desTweet(tweetMsg)
  }
  return tweets
}

func desTweet(msg *pb.Tweet) Tweet {
  if msg == nil {
    return Tweet{}
  }
  tweet := Tweet{
    ID:                   msg.Id,
    CreatedAt:            time.Unix(msg.CreatedAt, 0),
    Text:                 msg.Text,
    TextDisplayRange:     desIndices(msg.TextDisplayRange),
    Truncated:            msg.Truncated,
    Source:               msg.Source,
    User:                 desUser(msg.User),
    Quotes:               uint(msg.QuoteCount),
    Replies:              uint(msg.ReplyCount),
    Retweets:             uint(msg.RetweetCount),
    Likes:                uint(msg.FavoriteCount),
    CurrentUserLiked:     msg.Favorited,
    CurrentUserRetweeted: msg.Retweeted,
    CurrentUserRetweetID: desOptFixed64(msg.CurrentUserRetweetId),
    Hashtags:             desSymbols(msg.Hashtags),
    URLs:                 desURLs(msg.Urls),
    Mentions:             desMentions(msg.Mentions),
    Symbols:              desSymbols(msg.Symbols),
    Media:                desMediaItems(msg.Media),
    Polls:                desPolls(msg.Polls),
    PossiblySensitive:    msg.PossiblySensitive,
    FilterLevel:          msg.FilterLevel,
    Lang:                 msg.Lang,
    WithheldCopyright:    msg.WithheldCopyright,
    WithheldCounties:     msg.WithheldCountries,
    WithheldScope:        msg.WithheldScope,
  }
  if msg.RepliedTweet != nil {
    tweet.RepliedTo = &ReplyData{
      TweetID:    msg.RepliedTweet.ReplyToTweetId,
      UserID:     msg.RepliedTweet.ReplyToUserId,
      UserHandle: msg.RepliedTweet.ReplyToUserHandle,
    }
  }
  if msg.QuotedTweet != nil {
    tweet.Quoted = new(Tweet)
    *tweet.Quoted = desTweet(msg.QuotedTweet)
  }
  if msg.RepliedTweet != nil {
    tweet.Retweeted = new(Tweet)
    *tweet.Retweeted = desTweet(msg.RetweetedTweet)
  }
  return tweet
}

func desUser(msg *pb.User) User {
  if msg == nil {
    return User{}
  }
  return User{
    ID:                  msg.Id,
    Handle:              msg.Handle,
    DisplayName:         msg.DisplayName,
    CreatedAt:           time.Unix(msg.CreatedAt, 0),
    Bio:                 msg.Bio,
    URL:                 msg.Url,
    Location:            msg.Location,
    Protected:           msg.Protected,
    Verified:            msg.Verified,
    FollowerCount:       uint(msg.FollowerCount),
    FollowingCount:      uint(msg.FollowingCount),
    ListedCount:         uint(msg.ListedCount),
    FavouritesCount:     uint(msg.FavoritesCount),
    StatusesCount:       uint(msg.StatusesCount),
    ProfileBanner:       msg.ProfileBanner,
    ProfileImage:        msg.ProfileImage,
    DefaultProfile:      msg.DefaultProfile,
    DefaultProfileImage: msg.DefaultProfileImage,
    WithheldCountries:   msg.WithheldCountries,
    WithheldScope:       msg.WithheldScope,
    URLs:                desURLs(msg.UrlUrls),
    BioURLs:             desURLs(msg.BioUrls),
  }
}

func desIndices(msg *pb.Indices) Indices {
  if msg == nil {
    return Indices{}
  }
  return Indices{
    Start: uint(msg.Start),
    End:   uint(msg.End),
  }
}

func desURLs(msgs []*pb.URL) []URL {
  urls := make([]URL, len(msgs))
  for i, msg := range msgs {
    urls[i] = desURL(msg)
  }
  return urls
}

func desURL(msg *pb.URL) URL {
  if msg == nil {
    return URL{}
  }
  return URL{
    Indices:     desIndices(msg.Indices),
    TwitterURL:  msg.TwitterUrl,
    DisplayURL:  msg.DisplayUrl,
    ExpandedURL: msg.ExpandedUrl,
  }
}

func desSymbols(msgs []*pb.Symbol) []Symbol {
  symbols := make([]Symbol, len(msgs))
  for i, msg := range msgs {
    if msg != nil {
      symbols[i] = Symbol{
        Indices: desIndices(msg.Indices),
        Text:    msg.Text,
      }
    }
  }
  return symbols
}

func desMentions(msgs []*pb.Mention) []Mention {
  mentions := make([]Mention, len(msgs))
  for i, msg := range msgs {
    if msg != nil {
      mentions[i] = Mention{
        Indices:         desIndices(msg.Indices),
        UserID:          msg.UserId,
        UserHandle:      msg.Handle,
        UserDisplayName: msg.DisplayName,
      }
    }
  }
  return mentions
}

func desMediaItems(msgs []*pb.Media) []Media {
  media := make([]Media, len(msgs))
  for i, msg := range msgs {
    if msg != nil {
      media[i] = Media{
        URL:           desURL(msg.Url),
        ID:            msg.Id,
        Type:          msg.Type,
        MediaURL:      msg.MediaUrl,
        Alt:           msg.Alt,
        SourceTweetID: desOptFixed64(msg.SourceTweetId),
        Thumb:         desMediaSize(msg.Thumb),
        Small:         desMediaSize(msg.Small),
        Medium:        desMediaSize(msg.Medium),
        Large:         desMediaSize(msg.Large),
      }
    }
  }
  return media
}

func desMediaSize(msg *pb.Media_Size) MediaSize {
  if msg == nil {
    return MediaSize{}
  }
  return MediaSize{
    Width:  uint(msg.Width),
    Height: uint(msg.Height),
    Resize: msg.Resize,
  }
}

func desPolls(msgs []*pb.Poll) []Poll {
  polls := make([]Poll, len(msgs))
  for i, msg := range msgs {
    if msg != nil {
      polls[i] = Poll{
        EndTime:  time.Unix(msg.EndTime, 0),
        Duration: time.Minute * time.Duration(msg.DurationMinutes),
        Options:  make([]PollOption, len(msg.Options)),
      }
      for j, optMsg := range msg.Options {
        if optMsg != nil {
          polls[i].Options[j] = PollOption{
            Position: uint(optMsg.Position),
            Text:     optMsg.Text,
          }
        }
      }
    }
  }
  return polls
}

func desOptFixed64(msg *pb.OptFixed64) *uint64 {
  if msg != nil {
    val := new(uint64)
    *val = msg.Val
    return val
  }
  return nil
}

func desOptString(msg *pb.OptString) *string {
  if msg != nil {
    val := new(string)
    *val = msg.Val
    return val
  }
  return nil
}
